### ISSUE TITLE
unrar and plex transcoder not found on NVidia Shield Pro TV (2019 model)

### ISSUE THREAD

**Issue by janusloo** [pannal/Sub-Zero.bundle#783]
Hi there,

First of all I would like to thank you for all the work you have done on this plugin. It is really awesome and it saves me a lot of time manually downloading subtitles. 

I encountered 2 issues at the moment and could not find any solutions after looking at the issues here. From the DEBUG log I notieced that unrar is not found but it is in the path specified in the debug log and it already has exec bit on it:
`2022-02-27 17:54:33,761 (2eef9ad4f0) :  DEBUG (core:539) - Server version is 1.25.3.5575-f11334058
2022-02-27 17:54:33,764 (2eef8af4f0) :  DEBUG (services:438) - No shared code to load
2022-02-27 17:55:03,501 (2eeabc9548) :  DEBUG (config:350) - setting generic executable permissions for /storage/9C7F-CBCD/Android/data/com.plexapp.mediaserver.smb/Plex Media Server/Plug-ins/Sub-Zero.bundle/Contents/Libraries/Linux/aarch64/UnRAR/unrar
2022-02-27 17:55:03,519 (2eeabc9548) :  DEBUG (config:363) - custom check failed for: /storage/9C7F-CBCD/Android/data/com.plexapp.mediaserver.smb/Plex Media Server/Plug-ins/Sub-Zero.bundle/Contents/Libraries/Linux/aarch64/UnRAR/unrar
2022-02-27 17:55:03,544 (2eeabc9548) :  DEBUG (config:363) - custom check failed for: unrar
2022-02-27 17:55:03,546 (2eeabc9548) :  WARNING (config:373) - UnRAR not found
2022-02-27 17:55:03,572 (2eeabc9548) :  DEBUG (preferences:258) - Loaded preferences from DefaultPrefs.json
2022-02-27 17:55:03,595 (2eeabc9548) :  DEBUG (preferences:178) - Loaded the user preferences for com.plexapp.agents.subzero
2022-02-27 17:55:03,650 (2eeabc9548) :  INFO (datakit:93) - Loaded the dictionary file`

It is very strange. I have also reinstalled the plugin but the error still exists. Maybe you can shed some light on this why it should fail?

And the 2nd issue I have is related to plex transcoder not found when sub-zero tries to extract the SRT subtitle from the media:
`2022-02-27 18:01:11,741 (2ef253d4f0) :  DEBUG (logger:28) - subzero.video: Parsing video: Foundation (2021) - S01E01 - The Emperor’s Peace [ATVP WEBDL-720p][EAC3 5.1][h264]-CasStudio.mkv, hints: {'episode_title': u'The Emperor\u2019s Peace', 'type': 'episode', 'title': 'Foundation (2021)'}
2022-02-27 18:01:11,742 (2ef253d4f0) :  INFO (logger:38) - subliminal_patch.core: Determining basic video properties for 'Foundation (2021) - S01E01 - The Emperor\xe2\x80\x99s Peace [ATVP WEBDL-720p][EAC3 5.1][h264]-CasStudio.mkv' in '/storage/QNAPNAS/TV Series/Foundation (2021)/Season 1'
2022-02-27 18:01:11,961 (2ef253d4f0) :  DEBUG (logger:28) - subliminal_patch.core: GuessIt found: {
    "title": "Foundation (2021)", 
    "season": 1, 
    "episode": 1, 
    "episode_title": "The Emperor’s Peace", 
    "format": "WEB-DL", 
    "screen_size": "720p", 
    "audio_codec": "EAC3", 
    "audio_channels": "5.1", 
    "video_codec": "h264", 
    "release_group": "CasStudio", 
    "container": "mkv", 
    "type": "episode"
}
2022-02-27 18:01:12,079 (2ef253d4f0) :  INFO (scanning:128) - Found audio streams: en
2022-02-27 18:01:12,082 (2ef253d4f0) :  DEBUG (logger:28) - subzero.video: Determining existing subtitles for /storage/QNAPNAS/TV Series/Foundation (2021)/Season 1/Foundation (2021) - S01E01 - The Emperor’s Peace [ATVP WEBDL-720p][EAC3 5.1][h264]-CasStudio.mkv
2022-02-27 18:01:12,083 (2ef253d4f0) :  DEBUG (logger:28) - subliminal_patch.core: external subs: scanning path /storage/QNAPNAS/TV Series/Foundation (2021)/Season 1/Foundation (2021) - S01E01 - The Emperor’s Peace [ATVP WEBDL-720p][EAC3 5.1][h264]-CasStudio.mkv
2022-02-27 18:01:12,089 (2ef253d4f0) :  DEBUG (logger:28) - subliminal_patch.core: Found subtitles {}
2022-02-27 18:01:12,090 (2ef253d4f0) :  DEBUG (logger:28) - subliminal_patch.core: external subs: found {}
2022-02-27 18:01:12,091 (2ef253d4f0) :  WARNING (__init__:182) - Plex Transcoder not found, can't auto extract`

I have no idea why your plugin cannot find the plex transcoder and I also don't know where the location is of the transcoder on NVidia Shield TV Pro....

Can you please help me? Thank you! I got the info from the debug log. If you need more information pls let me know! Thank you!

**Comment by janusloo** [comment]
I found issue #460 regarding the Plex Transcoder location on Nvidia Shield but it is unfortunately not in the path indicated in the post. When I set plex_transcoder = "/data/user/0/com.plexapp.mediaserver.smb/Plex Transcoder" or plex_transcoder = "/data/user/0/com.plexapp.mediaserver.smb/Resources/Plex Transcoder" in config.py the plugin does not start after clicking Restart Plugin in Kitana. This is still not working despite fixing issue #460.

**Comment by janusloo** [comment]
I fixed the unrar not found error by commenting out the following codes in config.py:

` for exe in try_executables:
            rarfile.UNRAR_TOOL = exe
            rarfile.ORIG_UNRAR_TOOL = exe
            #if os.path.isfile(exe) and not os.access(exe, os.X_OK):
            #    st = os.stat(exe)
            #    try:
            #        Log.Debug("setting generic executable permissions for %s", exe)
                    # fixme: too broad?
            #        os.chmod(exe, st.st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
            #    except:
            #        Log.Debug("failed setting generic executable permissions for %s: %s", exe, traceback.format_exc())
            #        try:
            #            Log.Debug("setting executable permissions for %s", exe)
            #            os.chmod(exe, st.st_mode | stat.S_IEXEC)
            #        except:
            #            Log.Debug("failed setting executable permissions for %s: %s", exe, traceback.format_exc())
            #try:
            #    rarfile.custom_check([rarfile.UNRAR_TOOL], True)
            #except:
            #    Log.Debug("custom check failed for: %s", exe)
            #    continue
`

Not the most elegant way but it fixes my error at the moment.

**Comment by janusloo** [comment]
Hi @pannal can you please take a look at this? Thank you! I tried to root my 2019 Nvidia Shield TV Pro in the past few days but I could not find any instructions that work so I have no chance to look into the filesystem where the binary file is located.

**Comment by pannal** [comment]
Hmm, both strange issues. I've never used SZ on an android device so I don't have much experience there.

Seems like the unrar exe that it tries is the aarch64 one which is probably wrong.

You're probably better off trying bazarr instead of SZ. Development on SZ has basically halted now as the old Plex agents are deprecated and can be gone with any upcoming release, disabling the python functionality and thus any plugins such as SZ.

**Comment by janusloo** [comment]
Hi @pannal thank you for your reply. I can confirm that Nvidia Shield TV Pro 2019 model is 64 bits ARM, as confirmed by PMS log:

Processor: 4-core AArch64

So if I understood you well, you have no idea what the full path to Plex Transcoder is on Android, specifically NVidia Shield TV Pro?

I used Bazaar before but it never auto found any subtitles for me, SZ does a great job so far!

**Comment by pannal** [comment]
> So if I understood you well, you have no idea what the full path to Plex Transcoder is on Android, specifically NVidia Shield TV Pro?

Well I have had _some_ incling in the past as I've added multiple paths that are checked for existance. But that was a long time ago, and there have been many changes to Android file system handling.

If you can find out, and if you're able to execute, I could at least point you to how to add that path to the scan or even release a new dev build for you.

**Comment by janusloo** [comment]
Hi @pannal I did flash it with rooted developer's firmware and I used adb root to search all filesystems for plex transcoder, plex_transcoder, plex or transcoder (with find -iname) but could not find anything. Probably Plex has changed the transcoder name? I would like to help you with finding the correct transcoder. I even filed a ticket on plex forum in the nvidia section but until now I still got no reply from Plex support. It would be really cool to get the subtitle extraction to work.

Does subtitle extraction also work for PGS subtitles?
